### PR TITLE
Add fuzzing integration based on cargo-fuzz

### PR DESCRIPTION
**Pull request by 197g**
The initial corpus simply consists of all test files.

First ensure that [`cargo-fuzz`](https://github.com/rust-fuzz/cargo-fuzz) is installed. You may want to read its documentation but running is not complicated at all. In the top-level project directory:

```
cargo fuzz run reader
```

When it finds crashes (and it currently does quickly) then output files are written to
> `fuzz/artifacts/reader`

It is possible to replicate their runs by manually running the compiled fuzzing binary, providing them with a single argument point to the crashing input file.

The corpus (`fuzz/corpus/reader/`) is used to seed the automatic discovery of input test cases. It will be automatically updated based on unique code paths but is explicitely ignored by git. Newly discovered crashes can nevertheless be useful and you may want to forcibly add fixed test cases to the corpus, as they function as base test cases as well.

### CODE REVIEW

File: fuzz/fuzz_targets/reader.rs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+        let _ = item.node_name();
+        let _ = item.unit_address();
+        let _ = item.value();
+        let _ = item.value_str();
+        let _ = item.value_str_list(&mut large_buffer[..]);
+        let _ = item.value_u32_list(&mut large_buffer[..]);
+    }
+
+    for entry in reader.reserved_mem_entries() {
+        drop(entry);
```
Comment (ababo) [inline fuzz/fuzz_targets/reader.rs:25]: Why do we need `drop()` here?

File: fuzz/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+[dependencies.libfuzzer-sys]
+git = "https://github.com/rust-fuzz/libfuzzer-sys.git"
+
+# Prevent this from interfering with workspaces
+[workspace]
+members = ["."]
+
+[[bin]]
+name = "reader"
+path = "fuzz_targets/reader.rs"
```
Comment (ababo) [inline fuzz/Cargo.toml:22]: Can't we redirect `corpus` to `test_dtb` not to duplicate the files?

Comment (197g) [reply fuzz/Cargo.toml:22]: Maybe, and I looked it up for a while but didnt't a direct way to persistently configure `libfuzzer` so it didn't seem worth the benefit while maintaining the simple command line. The files are small by convention and not actually duplicated in git so it did not seem worth the extra hours.

File: fuzz/fuzz_targets/reader.rs (hunk 1 above)
Comment (197g) [reply fuzz/fuzz_targets/reader.rs:25]: Just to make it clear that this loop is empty by design. There is no inherent method to test here.

**Review by ababo — CHANGES_REQUESTED** [review]
Please remove the `corpus` directory, since it can be supplied directly via command line:

```bash
cargo fuzz run reader src/test_dtb
```
I would also ask to update `README.md` with instruction for fuzzy-testing (with mentioning `rustc-nightly` compiler).

**Comment by 197g** [comment]
> with mentioning rustc-nightly compiler

Did you need nightly to install `cargo-fuzz`, or for executing the fuzzer? I don't think it is necessary.

**Comment by ababo** [comment]
> Did you need nightly to install cargo-fuzz, or for executing the fuzzer?

Yes, it doesn't work with the stable compiler version.

**Comment by 197g** [comment]
Added a short note on this.

**Review by ababo — APPROVED** [review]

**Pull request merged**